src/main/java/com/couchbase/lite/support/action/Action.java

```diff
@@ -0,0 +1,342 @@
+     * @param backout A block that undoes the effect of the action; it will be called if a _later_
+     *                action fails, so that the system can be returned to the initial state.
+     * @param cleanup A block that performs any necessary cleanup after all actions have been
+     *                performed (e.g. deleting a temporary file.)
+     */
+    public void add(ActionBlock perform, ActionBlock backout, ActionBlock cleanup) {
+        peforms.add(perform != null ? perform : nullAction);
+        backouts.add(backout != null ? backout : nullAction);
+        cleanUps.add(cleanup != null ? cleanup : nullAction);
+    }
+
+    /**
+     * Adds an action as a step of this one. The action has two components (backout and cleanup
+     * are the same action), each optional.
+     * @param perform A block that tries to perform the action, or returns an error if it fails.
+     *                (If the block fails, it should clean up; the backOut will _not_ be called!)
+     * @param backoutOrCleanup A block that can be both backout and cleanup block.A backout block.
+     */
+    public void add(ActionBlock perform, ActionBlock backoutOrCleanup) {
+        add(perform, backoutOrCleanup, backoutOrCleanup);
+    }
+
+    /**
+     * Performs all the actions in order.
+     * If any action fails, backs out the previously performed actions in reverse order.
+     * If the actions succeeded, cleans them up in reverse order.
+     * The `lastError` property is set to the exception thrown by the failed perform block.
+     * The `failedStep` property is set to the index of the failed perform block.
+     * @throws ActionException
+     */
+    public void run() throws ActionException {
+        try {
+            perform();
+            try {
+                cleanup(); // Ignore exception
+            } catch (ActionException e) {}
+            lastError = null;
+        } catch (ActionException e) {
+            // (perform: has already backed out whatever it did)
+            lastError = e;
+            throw e;
+        }
+    }
+
+    @Override
+    public void perform() throws ActionException {
+        if (nextStep > 0)
+            throw new IllegalStateException("Actions have already been run");
+
+        failedStep = -1;
+        for (; nextStep < peforms.size(); ++nextStep) {
+            try {
+                doAction(peforms);
+            } catch (ActionException e) {
+                failedStep = nextStep;
+                if (nextStep > 0) {
+                    try {
+                        backout(); // Ignore error
+                    } catch (ActionException backoutEx) {}
+                }
+                throw e;
+            }
+        }
+    }
+
+    @Override
+    public void backout() throws ActionException {
+        if (nextStep == 0)
+            throw new IllegalStateException("Actions have not been run");
+
+        while(nextStep-- > 0) {
```

Comment:
I guess backout for current step is not executed. Is it expected?

If error occurs with current step, how does it recover it?